scripts/esrupgrades/esr.5.37-7.8.mysql.up.sql

```diff
@@ -1,1295 +1,1376 @@
 
-/* ==> mysql/000074_upgrade_users_v6.3.up.sql <== */
+	-- 'CREATE INDEX idx_posts_root_id_delete_at ON Posts(RootId, DeleteAt);'
+	DECLARE CreateIndexRootId BOOLEAN;
+	DECLARE CreateIndexRootIdQuery TEXT DEFAULT NULL;
 
-SET @preparedStatement = (SELECT IF(
-    (
-        SELECT COUNT(*) FROM INFORMATION_SCHEMA.COLUMNS
-        WHERE table_name = 'Users'
-        AND table_schema = DATABASE()
-        AND column_name = 'AcceptedTermsOfServiceId'
-    ) > 0,
-    'ALTER TABLE Users DROP COLUMN AcceptedTermsOfServiceId;',
-    'SELECT 1'
-));
-
-PREPARE alterIfExists FROM @preparedStatement;
-EXECUTE alterIfExists;
-DEALLOCATE PREPARE alterIfExists;
+	-- 'DROP INDEX idx_posts_root_id ON Posts;',
+	DECLARE DropIndex BOOLEAN;
+	DECLARE DropIndexQuery TEXT DEFAULT NULL;
 
-/* ==> mysql/000075_alter_upload_sessions_index.up.sql <== */
+	-- 'CREATE INDEX idx_posts_create_at_id on Posts(CreateAt, Id) LOCK=NONE;'
+	DECLARE CreateIndexCreateAt BOOLEAN;
+	DECLARE CreateIndexCreateAtQuery TEXT DEFAULT NULL;
+
+	SELECT COUNT(*) FROM INFORMATION_SCHEMA.COLUMNS
+		WHERE TABLE_NAME = 'Posts'
+		AND table_schema = DATABASE()
+		AND COLUMN_NAME = 'ParentId'
+		INTO DropParentId;
+
+	SELECT COUNT(*) FROM INFORMATION_SCHEMA.COLUMNS
+		WHERE table_name = 'Posts'
+		AND table_schema = DATABASE()
+		AND column_name = 'FileIds'
+		AND LOWER(column_type) != 'text'
+		INTO ModifyFileIds;
+
+	SELECT COUNT(*) FROM INFORMATION_SCHEMA.COLUMNS
+		WHERE table_name = 'Posts'
+		AND table_schema = DATABASE()
+		AND column_name = 'Props'
+		AND LOWER(column_type) != 'json'
+		INTO ModifyProps;
+
+	SELECT COUNT(*) = 0 FROM INFORMATION_SCHEMA.STATISTICS
+		WHERE table_name = 'Posts'
+		AND table_schema = DATABASE()
+		AND index_name = 'idx_posts_root_id_delete_at'
+		INTO CreateIndexRootId;
+
+	SELECT COUNT(*) FROM INFORMATION_SCHEMA.STATISTICS
+		WHERE table_name = 'Posts'
+		AND table_schema = DATABASE()
+		AND index_name = 'idx_posts_root_id'
+		INTO DropIndex;
+
+	SELECT COUNT(*) = 0 FROM INFORMATION_SCHEMA.STATISTICS
+		WHERE table_name = 'Posts'
+		AND table_schema = DATABASE()
+		AND index_name = 'idx_posts_create_at_id'
+		INTO CreateIndexCreateAt;
+
+	IF DropParentId THEN
+		SET DropParentIdQuery = 'DROP COLUMN ParentId';
+		UPDATE Posts SET RootId = ParentId WHERE RootId = '' AND RootId != ParentId;
+	END IF;
+
+	IF ModifyFileIds THEN
+		SET ModifyFileIdsQuery = 'MODIFY COLUMN FileIds text';
+	END IF;
+
+	IF ModifyProps THEN
+		SET ModifyPropsQuery = 'MODIFY COLUMN Props JSON';
+	END IF;
+
+	IF CreateIndexRootId THEN
+		SET CreateIndexRootIdQuery = 'ADD INDEX idx_posts_root_id_delete_at (RootId, DeleteAt)';
+	END IF;
+
+	IF DropIndex THEN
+		SET DropIndexQuery = 'DROP INDEX idx_posts_root_id';
+	END IF;
+
+	IF CreateIndexCreateAt THEN
+		SET CreateIndexCreateAtQuery = 'ADD INDEX idx_posts_create_at_id (CreateAt, Id)';
+	END IF;
+
+	IF DropParentId OR ModifyFileIds OR ModifyProps OR CreateIndexRootId OR DropIndex OR CreateIndexCreateAt THEN
```

Comment:
In a case like this, I wonder if it would be easier to:
```
SET @query = CONTACT_WS(', ', Conditional1, Conditional2, Conditional3, ...)
IF @query <> '' THEN
    SET @fullQuery = CONTACT('ALTER TABLE ...', @query)
    ...
END IF
```

With the sole advantage being to eliminate the duplicate lists of conditionals (also reducing the chance we miss one).

Reply:
Ooooh, that's clever, I like it!

Reply:
Let's discuss this change in the other thread as well